hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieBaseParquetWriter.java

```diff
@@ -65,14 +63,18 @@ public HoodieBaseParquetWriter(Path file,
   }
 
   public boolean canWrite() {
-    // TODO we can actually do evaluation more accurately:
-    //      if we cache last data size check, since we account for how many records
-    //      were written we can accurately project avg record size, and therefore
-    //      estimate how many more records we can write before cut off
-    if (lastCachedDataSize == -1 || getWrittenRecordCount() % WRITTEN_RECORDS_THRESHOLD_FOR_FILE_SIZE_CHECK == 0) {
-      lastCachedDataSize = getDataSize();
+    if (getWrittenRecordCount() >= recordNumForNextCheck) {
+      long dataSize = getDataSize();
+      long avgRecordSize = dataSize / getWrittenRecordCount();
+      // Follow the parquet block size check logic here, return false
+      // if it is within ~2 records of the limit
+      if (dataSize > (maxFileSize - avgRecordSize * 2)) {
+        return false;
+      }
+      // Do check it in the halfway
+      recordNumForNextCheck = (recordNumForNextCheck + maxFileSize / avgRecordSize) / 2;
```

Comment:
There is also some upper-threshold protection logic:

```java
    recordCountForNextMemCheck = min(
            max(props.getMinRowCountForPageSizeCheck(),
              (recordCount + (long)(nextRowGroupSize / ((float)recordSize))) / 2), // will check halfway
            recordCount + props.getMaxRowCountForPageSizeCheck() // will not look more than max records ahead
            );
```

Any reason we missed it here? For huge `maxFileSize` or very small `avgRecordSize`, the `recordNumForNextCheck` can be very large, that would cause the final file size to exceed the disired bar significantly.

Reply:
In the first place, I thought this issue only happens if `avgRecordSize` can vary greatly(large variance), and in such case `props.getMaxRowCountForPageSizeCheck()` doesn't do much help, because it needs users well understanding their data.

But I think you're right, let me add these two configures `getMinRowCountForPageSizeCheck`, `getMaxRowCountForPageSizeCheck`, at least gives some users ability to adjust it.